platform/website/src/components/auth/auth-dialog.svelte

```diff
@@ -48,262 +50,99 @@
 	let turnstileToken = "";
 	let loggingIn = false;
 
-	let username = newField({
-		id: "username",
-		type: "text",
-		label: "Username",
-		autoComplete: "username",
-		extra: {
-			timeout: undefined as NodeJS.Timeout | undefined,
-		},
-		reload() {
-			username.touched = username.value.length > 0;
-
-			if ($authDialog === AuthDialog.Login) {
-				username.message = "";
-				username.status = "";
-			}
-
-			if (username.touched) {
-				username.update(username.value);
-			}
-		},
-		update(value) {
-			username.value = value;
+	$: $authDialog, resetAllFields();
 
-			// We clear the timeout so that we dont send a request to the server on every character change.
-			clearTimeout(username.extra.timeout);
+	let usernameValue: string;
+	let usernameStatus: FieldStatus;
+	let usernameValidationTimeout: number | NodeJS.Timeout;
+	async function usernameValidate(v: string) {
+		globalMessage = "";
+		if ($authDialog === AuthDialog.Login) {
+			return { type: FieldStatusType.Success };
+		}
 
-			// When we are in login mode, we dont need to validate the username.
-			if ($authDialog !== AuthDialog.Register) {
-				return;
-			}
+		clearTimeout(usernameValidationTimeout);
 
-			const valid = z
-				.string()
-				.min(3, "Minimum of 3 characters")
-				.max(20, "Maximum of 20 characters")
-				.regex(/^[a-zA-Z0-9_]*$/, "Username can only contain letters, numbers, and underscores")
-				.safeParse(username.value);
-			if (!valid.success) {
-				username.status = "error";
-				username.message = valid.error.issues[0].message;
-				return;
-			}
+		const valid = z
+			.string()
+			.min(3, "Minimum of 3 characters")
+			.max(20, "Maximum of 20 characters")
+			.regex(/^[a-zA-Z0-9_]*$/, "Username can only contain letters, numbers, and underscores")
+			.safeParse(v);
+		if (!valid.success) {
+			return { type: FieldStatusType.Error, message: valid.error.issues[0].message };
+		}
 
-			username.message = "Hold on while we check if this username is available...";
-			username.status = "loading";
-
-			// This is known as a debouncer.
-			// We dont want to send a request to the server on every character change.
-			// Instead we wait 200ms after the user stops typing before sending the request.
-			username.extra.timeout = setTimeout(async () => {
-				const result = await client
-					.query(
-						graphql(`
-							query CheckUsername($username: String!) {
-								user {
-									user: byUsername(username: $username) {
-										id
-									}
-								}
+		usernameStatus = {
+			type: FieldStatusType.Loading,
+			message: "Hold on while we check if this username is available...",
+		};
+
+		await new Promise((resolve) => {
+			usernameValidationTimeout = setTimeout(resolve, 500);
```

Comment:
We still need to fix this, by exposing the `reject` part of the promise outside, and making it so when you cancel the timeout the reject gets called.